### ISSUE TITLE
Plus Server Error with 3D Slicer

### ISSUE THREAD

**Issue by hripat** [PlusToolkit/PlusLib#619]
Hello,
I am trying to use Slicer with the Plus Server connected so that I can use the transforms function for the Ultrasound images. However, I am getting the error show in the image. I have installed all four SlicerIGT related extensions in Slicer but I’m not sure what’s causing this error.

The error says:
|ERROR|022.003000| Cannot connect to the server (127.0.0.1:18946).| in E:\D\PSNP64b\PlusLib\src\PlusDataCollection\OpenIGTLink\vtkPlusOpenIGTLinkDevice.cxx(152)
|ERROR|022.311000| Socket error in device TrackerDevice: failed to receive OpenIGTLink transforms. Attempt to reconnect.| in E:\D\PSNP64b\PlusLib\src\PlusDataCollection\OpenIGTLink\vtkPlusOpenIGTLinkDevice.cxx(241)

![image](https://user-images.githubusercontent.com/56734075/67109333-48cc8300-f19e-11e9-8f76-d0fe0493df01.png)

**Comment by Sunderlandkyl** [comment]
Can you show the OpenIGTLink connections in Slicer?

**Comment by hripat** [comment]
What are the OpenIGTLink connections? I've installed all 3 of the IGT extensions in Slicer along with Jupyter, are those the connections?

The server worked fine before I installed another version of Slicer, could it be a problem with that? I've also modified the mml file of the Slicer because I wanted to input a new STL file into Slicer.

**Comment by Sunderlandkyl** [comment]
Which version of Slicer are you using?

**Comment by hripat** [comment]
I'm using Slicer 4.11 - the one from October 1st/2nd

**Comment by Sunderlandkyl** [comment]
I just tested the ultrasound simulator with the Oct 7th version of Slicer + most recent stable release of Plus, and it worked.

Can you upload your config file?

**Comment by hripat** [comment]
Is the config file the one under the Plus library- "PlusDeviceSet_Server_SimulatedUltrasound.xml?" 

I don't have access to the desktop that I am using these files on right now. I will have to get back to you on Monday.

Should I also upload the mml file that I modified in Slicer?

**Comment by Sunderlandkyl** [comment]
PlusDeviceSet_Server_SimulatedUltrasound_3DSlicer.xml
Sure, if you upload the mrml file as well that would be helpful.

**Comment by hripat** [comment]
Hi,
I'm attaching the config file and the mrml file for Plus and slicer in the zipped folder below.

[Plus Server Error Files.zip](https://github.com/PlusToolkit/PlusLib/files/3752001/Plus.Server.Error.Files.zip)

**Comment by Sunderlandkyl** [comment]
Can you save the full Slicer scene as an mrb?

**Comment by hripat** [comment]
Hi,

I've added the bundled MRB file to the zipped file below as well.
[Slicer and PLUS Error.zip](https://github.com/PlusToolkit/PlusLib/files/3752126/Slicer.and.PLUS.Error.zip)

**Comment by Sunderlandkyl** [comment]
What version of Slicer were you using when it worked before? Slicer 4.8.1?

**Comment by hripat** [comment]
I was using 4.10.2. But I did also modify the mrml file and the config file to remove the needle and replace the STL file. I also changed the gel block reference as well.

**Comment by Sunderlandkyl** [comment]
Did you modify the mrml scene directly with a text editor, or just changed the scene in Slicer?

**Comment by hripat** [comment]
I modified it using Wordpad.

**Comment by Sunderlandkyl** [comment]
I see. I think that the modifications might be part of the problem. There were some strange things going on when I loaded the mrb into Slicer, such as missing node references, etc, but it seemed to work.

![image](https://user-images.githubusercontent.com/9222709/67233765-db288d00-f411-11e9-96b9-4ec9757953fb.png)

Can you try loading the unmodified PlusDeviceSet_Server_SimulatedUltrasound_3DSlicer.xml and mrb that come with Plus and check if that works?

**Comment by hripat** [comment]
Yes, the transforms work on the original xml file works. I was looking at my modified xml file and I was trying to delete the needle from the xml file because I don't need it for my work. However, in the Data Sources areas and other places, the "needle" object is still being referenced. Should I try removing those as well?

**Comment by Sunderlandkyl** [comment]
I would load the original mrb and modify the scene within Slicer to set up the scene the way that you want. Editing the scene with a text editor is probably not a good idea since it may cause issues with missing nodes, among other things.

**Comment by hripat** [comment]
How would I modify the scene in Slicer? I need to put in a new 3D object of the femur instead of the rod in place right now and would also like to remove the needle. Are there any tutorials for this?

**Comment by Sunderlandkyl** [comment]
You can drag and drop your models into Slicer.
If you have any questions about how to set up a specific element I'd be glad to help.

**Comment by hripat** [comment]
I was able to drop the femur STL into Slicer however, how would I get rid of the rod and the needle in the scene? I am not able to see a good ultrasound image of the femur while using the transforms.
![image](https://user-images.githubusercontent.com/56734075/67236994-55f4a680-f418-11e9-9516-b2ed9c962f48.png)

**Comment by Sunderlandkyl** [comment]
You can delete the rod and needle models in Slicer (Data module -> Right click ->Delete), or you can just hide them with the eye icon.

Now you can change the vessel model in the Plus config file to use femur.stl like in the config file you sent me.

**Comment by hripat** [comment]
I changed the PlusDeviceSet_Server_SimulatedUltrasound_3DSlicer.xml file to have the femur.stl file instead of the vessel but when I opened the SimulatedUltrasound_Scene.mrb file, the femur didn't show up. The femur file in the same folder as well. Do I have to unzip the file and change the mrml file as well?

**Comment by hripat** [comment]
This is what's showing up in Slicer. The simulated ultrasound seems to be of the femur but the rod is still visible and I can't see the femur.
![image](https://user-images.githubusercontent.com/56734075/67239020-3e1f2180-f41c-11e9-83c5-f8416d7bea37.png)

**Comment by Sunderlandkyl** [comment]
The mrb and Plus config file are separate. You need to save and use the mrb that you created by dragging and dropping the stl into Slicer.

The mesh is not sent between Plus and Slicer, so you have to add it to both individually.

**Comment by hripat** [comment]
Oh okay, thank you that worked! I appreciate all your help!

**Comment by Sunderlandkyl** [comment]
You're welcome! Glad to help!
